models/ed/R/model2netcdf.ED2.R

```diff
@@ -118,497 +226,661 @@ model2netcdf.ED2 <- function(outdir, sitelat, sitelon, start_date, end_date) {
+    patch.area <- ed.dat$AREA[[i]]    # m2  -- one entry per patch
+    pacoN      <- ed.dat$PACO_N[[i]]  # number of cohorts per patch
+    patch.area <- rep(patch.area, pacoN)  # patch areas, repped out to one entry per cohort
+    
+    # Now can get number of plants per cohort, which will be used for weighting. Note that area may have been (often/always is?) a proportion of total site area, rather than an absolute measure. In which case this nplant is a tiny and meaningless number in terms of actual number of plants. But that doesn't matter for weighting purposes. 
+    nplant <- plant.dens * patch.area
+    
+    # Get index of DBH bin each cohort belongs to      
+    dbh.bin <- sapply(dbh, function(x) which.min(x>c(dbh.breaks,Inf))) - 1
+    
+    
+    # For each PFT x DBH bin, average variables of interest, weighting by # of plants. Not all ED cohort variables are in per-plant units. This code would not be applicable to them without modification.
+    # However, it does handle two special cases. For NPLANT, it performs no weighting, but simply sums over cohorts in the PFT x DBH bin. For MMEAN_MORT_RATE_CO, it first sums over columns representing different mortality types first, then proceeds with weighting. 
+    for(j in 1:ndbh) {
+      for(k in 1:npft) {
+        ind <- (dbh.bin==j) & (pft == pfts[k])
+        
+        if(any(ind)) {
+          for(varname in varnames) {
+            if(varname == "NPLANT") {
+              # Return the total number of plants in the bin
+              out$NPLANT[i,j,k] <- sum(nplant[ind])
+            } else if(varname == "MMEAN_MORT_RATE_CO") {
+              # Sum over all columns 
+              mort = apply(ed.dat$MMEAN_MORT_RATE_CO[[i]][ind,, drop=F], 1, sum, na.rm=T)
+              out$MMEAN_MORT_RATE_CO[i,j,k] <- sum(mort * nplant[ind]) / sum(nplant[ind])
+            } else {
+              # For all others, just get mean weighted by nplant
+              out[[varname]][i,j,k] <- sum(ed.dat[[varname]][[i]][ind] * nplant[ind]) / sum(nplant[ind])
+            }
+            dimnames(out[[varname]]) <- list(months = times[ysel], dbhLowBound=dbh.breaks, pft=pft.names)
+          }
+        }
+      }
+    }
+  }
+  
+  out <- lapply(out, function(o) aperm(o, c(2,3,1)))
+  out$PFT <- pfts # will write this to the .nc file
+  
+  return(out)
+  
+} # read_E_files
 
-  }  ## end year loop
-
-} # model2netcdf.ED2
 ##-------------------------------------------------------------------------------------------------#
+
+# Function for put -E- values to nc_var list
+
+put_E_values <- function(yr, nc_var, out, begins, ends, pft.names, dbh.breaks, ...){
+  
+  s <- length(nc_var)
+  
+  data(pftmapping, package = "PEcAn.ED2")
+  pfts <- sapply(pft.names, function(x) pftmapping$ED[pftmapping$PEcAn == x]) 
+  
+  # ----- fill list
+  
+  t <- ncdf4::ncdim_def(name = "dtime", units = paste0("days since ", yr, "-01-01 00:00:00"), 
```

Comment:
@mdietze how should I define the time dimension for monthly outputs? if we have both T and E files, their time steps are different..should we align their time steps (writing half hourly DBH sounds weird), but then the alternative would be defining two time dimensions, what do you think?

Reply:
Define two time dimensions

Reply:
what should those time dimension names be? currently one is `time` the other is `dtime`

Reply:
Not 100% sure. Maybe ```month```?

Here's some things to take a look at, but neither is a definitive answer
http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch04s04.html
http://www.unidata.ucar.edu/software/netcdf/docs/BestPractices.html